### PR TITLE
feat(checklist): typed ChecklistService + reserved wiki.* namespace + funnel (#984)

### LINKED ISSUE #984: Checklist data model refactor: reserved metadata namespace + ChecklistService + single mutation path
## Summary

Refactor the checklist data model to:

1. Move CalDAV/sync bookkeeping (`updated_at`, `created_at`, `completed_at`, `completed_by`, `automated`, `sync_token`, tombstones) into a **reserved `agent.checklists.*` namespace** that external frontmatter tools cannot write to.
2. Add a dedicated **`ChecklistService` MCP/proto surface** for item-level mutation (add, update, toggle, delete, reorder).
3. Funnel **all** checklist mutations — web UI handlers, MCP tools, future CalDAV handlers — through a **single internal mutator path** (`checklistMutator`) that performs diff-on-write metadata bookkeeping atomically.
4. Migrate existing pages onto the new model.
5. Audit agent prompts and switch them off `MergeFrontmatter`-based read-modify-write.

This is a hard prerequisite for the CalDAV bridge (#983), but is independently valuable: today's `help_macro_checklist` literally documents the read-modify-write pattern as a footgun. After this lands, that footgun is gone.

## Why

The current model:

- Stores checklist data in user-mutable frontmatter at `checklists.<name>.items[]`.
- Exposes mutation only via the generic `MergeFrontmatter` / `ReplaceFrontmatter` tools, which require callers to read the entire items array, modify it, and write it back wholesale. The "important notes" section of `help_macro_checklist` warns about this in capital letters.
- Has no per-item identity, no per-item timestamps, no attribution data.
- Provides no mechanism to ensure that future write paths (a new agent, a new MCP tool, a CalDAV handler) advance the metadata clients depend on.

The fragility shows up the moment you try to bolt on sync (#983): every mutation path must remember to bump `updated_at`, or CalDAV clients hold stale ETags and miss agent-driven changes. "Remember to bump" is exactly the kind of constraint that breaks silently.

The fix is structural: make the bump impossible to skip by routing every mutation through one internal funnel and putting the bump in that funnel.

## Two-Tier Schema

**User-mutable** (`checklists.<name>.items[]`) — what `MergeFrontmatter` accepts (subject to Open Question #1) and what `ChecklistService` reads/writes through:

```toml
[[checklists.grocery-list.items]]
uid = "ckl-9f8a-..."           # immutable; rejected if changed on existing item
text = "Buy milk"
checked = false
tags = ["urgent"]
sort_order = 3000              # int64, sparse (multiples of 1000)
description = "the brand Kirsten likes"   # optional
due = "2026-04-30T17:00:00Z"   # optional, RFC3339
alarm = "..."                   # optional; encoded VALARM payload
```

**Wiki-managed** (`agent.checklists.<name>.*`) — reserved, external writes silently stripped (mirrors `agent.schedules.*` policy):

```toml
[agent.checklists.grocery-list]
sync_token = 47                 # monotonic counter, advances on any item add/edit/delete

[agent.checklists.grocery-list.items.ckl-9f8a---]
created_at   = "2026-04-25T13:00:00Z"
updated_at   = "2026-04-25T17:14:00Z"
completed_at = "2026-04-25T17:14:00Z"
completed_by = "sean"
automated    = false

[[agent.checklists.grocery-list.tombstones]]
uid          = "ckl-old-deleted-..."
deleted_at   = "2026-04-24T08:00:00Z"
gc_after     = "2026-05-01T08:00:00Z"
```

## ChecklistService API

New proto service. Methods:

| Method | Purpose |
|---|---|
| `AddItem(page, checklist, text, tags?, description?, due?, sort_order?)` | Create item; wiki generates `uid`, `created_at`, `updated_at`. |
| `UpdateItem(page, checklist, uid, fields_to_change)` | Mutate any user-mutable fields except `uid`. |
| `ToggleItem(page, checklist, uid)` | Sets `checked`. Transitions `completed_at`/`completed_by` based on direction. |
| `DeleteItem(page, checklist, uid)` | Removes item; tombstone written. |
| `ReorderItem(page, checklist, uid, new_sort_order)` | Updates `sort_order`; only re-densifies adjacent values when collisions appear. |
| `ListItems(page, checklist)` | Returns user fields + wiki-managed metadata for each item. |
| `GetChecklists(page)` | Enumerates all checklists on a page (used by CalDAV PROPFIND). |

All methods use the auth principal as the source of truth for `completed_by` / `automated` derivation — never accepted from input.

ETag-style optimistic concurrency: `UpdateItem` / `ToggleItem` / `DeleteItem` accept an optional `expected_updated_at`; mismatch returns `FailedPrecondition` (the wiki equivalent of CalDAV's `412`).

## The `checklistMutator` Funnel

Internal package. **Every** mutation to checklist data — without exception — goes through:

```go
checklistMutator.Apply(ctx, page, op) → (newState, error)
```

Callers (all internal):
- `ChecklistService` MCP handlers
- Web UI form handlers (the existing checklist HTTP endpoints)
- CalDAV `PUT` / `DELETE` handlers (#983)
- The migration job (one-shot)

The mutator does, atomically:

1. Load current page frontmatter.
2. Compute the after-state for the relevant `checklists.<name>` subtree.
3. Diff before vs after by `uid`:
   - **New uid (or wiki-injected for unsourced item)** → set `created_at = updated_at = now()`.
   - **Existing uid, content changed** → set `updated_at = now()`.
   - **`checked` flipped false→true** → also set `completed_at = now()`, `completed_by = principal`.
   - **`checked` flipped true→false** → clear `completed_at`/`completed_by`.
   - **uid missing in after-state** → tombstone (see Open Question #2).
4. Set `automated` from `ctx.Principal().IsAgent()` — the wiki's identity subsystem is the single source of truth; the funnel never reaches for Tailscale tags / MCP context / CLI flags directly. See [Identity Integration](#identity-integration) below.
5. Bump `sync_token` for the collection if anything changed.
6. Write user-mutable + wiki-managed atomically. Single round-trip to the page store.

External writes targeting `agent.checklists.*` are silently stripped at the frontmatter-tool layer (mirrors `agent.schedules.*` enforcement).

## Identity Integration

`completed_by` and `automated` are derived from the wiki's existing identity subsystem — the funnel never reaches into protocol-layer details (Tailscale tags, MCP session metadata, CLI flags) directly. Identity normalizes across entry points:

| Entry point | How `principal.IsAgent()` is determined |
|---|---|
| Tailscale-authenticated HTTP/gRPC (web UI, CalDAV via #983, MCP over network) | Tailscale identity → wiki-config mapping (tagged device or configured agent allow-list → true; otherwise false) |
| MCP via `claude-agent-acp` (scheduled-agent fires, dorium-* sessions, headless chat pool) | ACP session type signals agent → true |
| `wiki-cli` invocations | **Defaults to `IsAgent() = true`** — wiki-cli is overwhelmingly invoked by automation (scheduled tasks, heartbeat, ad-hoc agent scripts). Humans running it interactively set `WIKI_CLI_HUMAN=1` env var (or `--as-human` flag) to opt out. Safer default: a human forgetting the flag gets their action mis-attributed as automated (informational, non-destructive); the alternative would be agents falsely attributed to humans (misleading audit / reflection data). |
| Internal wiki callers (migration job, scheduled-agent summary writer, etc.) | Synthetic principal `system` with `IsAgent() = true` |

If the identity subsystem doesn't yet expose `IsAgent()`, this issue requires extending it. The principal struct should grow whatever fields the funnel needs (`Name()`, `IsAgent()`, possibly `Source()` for telemetry), and each entry point's auth handler populates the principal correctly.

`completed_by` records `principal.Name()`. The funnel never accepts `completed_by` from input.

## Migration

**Eager, via the wiki's existing job infrastructure** (same pattern used by `PageImportService` and the inventory audit). On first startup post-upgrade, register a `migrate-checklists` job that walks every page, migrates its checklists, and writes results to a status page.

For each `checklists.<name>` found on the page:

1. Generate `uid` for any item lacking one.
2. Backfill `sort_order` from array position × 1000 if absent.
3. Populate `agent.checklists.<name>.items.<uid>.{created_at, updated_at}` using the page's last-modified time as best-available approximation.
4. Stamp `agent.checklists.<name>.migrated_data_model = true`.

Status reporting: emit a `checklist_migration_report` page (or append to a system audit page) summarizing pages migrated, items touched, anomalies (e.g. items with malformed structure that couldn't be auto-fixed). Mirrors the inventory audit pattern.

Re-runnable: the job is idempotent — pages already stamped `migrated_data_model = true` are skipped on subsequent runs. `wiki-cli` exposes a manual trigger for re-running after a fix.

> **Note for #982**: the `migrated_tags_syntax` migration described in #982 should use the same job infrastructure, not the lazy on-read approach previously mentioned there. Patch #982 to align.

## Agent Prompt Audit

Grep scope (all known places agents touch checklists):

- `~/agent/skills/` — heartbeat, daily-reflection, weekly-dinner-planner, wiki-confirmations, simple-wiki-overseer, project-template, claude-session-explorer, etc.
- Wiki scheduled-agent prompts (the `agent.schedules.*` configs across pages).
- `CLAUDE.md` and `~/agent/config/heartbeat.md`.

Replace `MergeFrontmatter`-based read-modify-write of `checklists.*` with `ChecklistService` calls. Each replaced call site gets a corresponding test (or at least a validation that the prompt parses and runs without falling back to `MergeFrontmatter`).

`help_macro_checklist` rewritten to document the new API surface and **stop** documenting the old footgun. Ships via #980.

## Open Questions

**1. ~~Should `MergeFrontmatter` reject writes to `checklists.*.items` entirely?~~ — DECIDED: yes, reject.**

By precedent with `agent.*` (per `help_scheduled_agents`): generic frontmatter tools (`MergeFrontmatter`, `ReplaceFrontmatter`, `RemoveKeyAtPath`) reject writes touching reserved namespaces, and dedicated services are the only entry point. We extend that uniform policy: **`checklists.*` becomes reserved**, and `ChecklistService` is the only way to mutate it.

Generic frontmatter tools rejecting writes to:
- `agent.*` (existing)
- `checklists.*` (new)

`ChecklistService` accepts user-mutable fields, manages metadata internally, never accepts wiki-managed fields from input.

**Forces the agent prompt audit to land in the same release as this issue** — no deprecation runway. Audit scope is finite: heartbeat, daily-reflection, weekly-dinner-planner, wiki-confirmations, simple-wiki-overseer, project-template, daily-reflection. All must be migrated before the rejection lands.

See ADR draft below.

**2. Tombstones for deleted items.**

CalDAV `sync-collection` REPORT must report deletions to slow-syncing clients. Recommend: 7-day tombstones in `agent.checklists.<n>.tombstones[]`, GC'd by a startup task or background sweeper.

Confirming: 7 days OK? Other duration?

**3. Strict uid preservation on `MergeFrontmatter` items writes.**

If we accept items writes via `MergeFrontmatter` (Open Question 1 = b), and an agent sends an items array where existing items have lost their uids:

- **(a) Reject** the write with `InvalidArgument`. Forces correctness.
- **(b) Best-effort match by text + position** to recover uids. Fragile but tolerant.

Recommendation: **(a)** reject. Agents must read items via `ListItems` first (they get uids), then write back with uids preserved. (Or just use `ChecklistService` and avoid the issue entirely.)

**4. Migration timing.**

- **(a) Lazy** — on first read of each page post-upgrade. Distributed cost, consistent with `migrated_tags_syntax`.
- **(b) Eager** — startup task walks all pages, migrates, then serves. One slow boot, deterministic state.

Recommendation: **(a)** lazy. Faster restarts, consistent precedent.

## Acceptance

- [ ] `agent.checklists.*` namespace reserved; external writes silently stripped (test).
- [ ] `ChecklistService` proto + MCP handlers + wiki-cli coverage.
- [ ] `checklistMutator` package with diff-on-write, attribution derivation, atomic write. Single funnel — all mutation callers go through it.
- [ ] Web UI checklist endpoints refactored to call `checklistMutator` instead of mutating frontmatter directly.
- [ ] Migration job (lazy, with `migrated_data_model` marker).
- [ ] Agent prompt audit completed; all known mutation call sites switched to `ChecklistService`.
- [ ] `help_macro_checklist` rewritten for the new API; ships via #980.
- [ ] ADR `docs/adr/0001-reserved-namespaces-and-dedicated-services.md` committed alongside the implementation.
- [ ] Tests:
  - Adding an item via `ChecklistService` populates `created_at`/`updated_at`/automated correctly per principal.
  - Toggling check sets/clears `completed_at`/`completed_by`.
  - Concurrent `UpdateItem` calls with stale `expected_updated_at` return `FailedPrecondition`.
  - Deleting an item creates a tombstone; tombstone is GC'd after the grace period.
  - External `MergeFrontmatter`/`ReplaceFrontmatter`/`RemoveKeyAtPath` write touching `checklists.*` or `agent.checklists.*` is rejected with `InvalidArgument` naming `ChecklistService`.
  - Migration is idempotent: running twice produces the same state, no spurious bumps.
  - `sync_token` advances exactly once per logical mutation, not per write call.
  - Agent (automated principal) writes set `automated = true` without input being able to override it.

## Dependencies & Coordination

- **Blocks #983** (CalDAV bridge) — the bridge consumes `agent.checklists.*` for ETags, `sync_token`, attribution, tombstones.
- **Coordinates with #980** — `help_macro_checklist` rewrite ships embedded.
- **Coordinates with #982** — the new `ChecklistService` honors the `#tag` syntax (the description hashtag parsing from #983 also lands here).
- **Independent of #981** — no required interaction with page-hashtag indexing.

## ADR

This issue establishes a pattern that earlier work (`agent.schedules.*`, `agent.chat_context.*`) implicitly used but never formalized. Land the following ADR alongside the implementation, at `docs/adr/0001-reserved-namespaces-and-dedicated-services.md` (create the `docs/adr/` directory if it doesn't exist):

````markdown
# ADR 0001: Reserved Frontmatter Namespaces and Dedicated Services

## Status

Accepted — 2026-04-25

## Context

The wiki stores structured data in TOML frontmatter at the top of each page. Most fields are "loose" — `title`, `description`, ad-hoc keys, simple structured config like `[blog]` or `[inventory]` — and are happily edited by users, agents, and tools using the generic frontmatter API (`MergeFrontmatter`, `ReplaceFrontmatter`, `RemoveKeyAtPath`).

Some structured data, however, has **non-trivial invariants** the wiki itself enforces:

- `agent.schedules.*` — cron expressions must validate; status fields (`last_run`, `last_status`, `last_error_message`, `last_duration_seconds`) are populated by the wiki on each fire and must not be writable from outside.
- `agent.chat_context.*` — `last_updated` is server-stamped; `background_activity` is a rolling log written by the wiki when scheduled fires terminate.
- `checklists.*` (this ADR) — items have stable identity (`uid`); per-item `created_at`/`updated_at`/`completed_at`/`completed_by`/`automated` and per-collection `sync_token` must advance correctly on every mutation for CalDAV (#983) and reflection use cases to work.

When such data was reachable through generic frontmatter tools, two problems arose:

1. **"Remember to bump" hazards.** Every mutation path had to remember to advance the wiki-managed metadata. Adding a new mutation path silently broke downstream consumers.
2. **Spoofing risk.** An external write could set `completed_by = "sean"` to fake human attribution, or set a stale `updated_at` to confuse sync clients.

We addressed `agent.schedules.*` and `agent.chat_context.*` ad-hoc by reserving the `agent.*` namespace and routing through `AgentMetadataService`. We are now extending the same approach to `checklists.*`.

## Decision

Establish a uniform policy for any frontmatter subtree with non-trivial invariants:

1. **Reserved namespace.** Generic frontmatter tools (`MergeFrontmatter`, `ReplaceFrontmatter`, `RemoveKeyAtPath`) reject writes that touch the reserved subtree, returning `InvalidArgument` with a message naming the dedicated service.
2. **Dedicated service.** A typed gRPC/MCP service is the only mutation entry point. It accepts user-mutable fields, derives wiki-managed fields from authoritative sources (auth principal, server clock, diff state), and writes both subtrees atomically.
3. **Internal mutator funnel.** Inside the wiki process, every caller of the dedicated service — gRPC handlers, web UI form handlers, future protocol bridges (e.g. CalDAV in #983) — converges on a single internal mutator function that performs the diff, attribution derivation, and atomic write. The funnel is the only code path that can write the reserved subtree.
4. **No silent-strip on dedicated-service input.** When the dedicated service accepts a payload that includes wiki-managed fields, it strips them silently — input never overrides server-derived values. (This matches how `AgentMetadataService.UpsertSchedule` strips `last_run` etc.)
5. **Identity-derived attribution.** Wiki-managed fields that record "who did this" (`completed_by`) and "was this automated" (`automated`) are derived from the wiki's identity subsystem, not from input or protocol-layer details. The funnel calls `ctx.Principal().Name()` and `ctx.Principal().IsAgent()`. Each entry-point's auth handler populates the principal; the funnel stays protocol-agnostic.

Currently reserved namespaces under this policy:

| Namespace | Service | Funnel |
|---|---|---|
| `agent.schedules.*` | `AgentMetadataService.{Upsert,Delete,List}Schedule` | (existing) |
| `agent.chat_context.*` | `AgentMetadataService.{Get,Update}ChatContext`, `AppendBackgroundActivitySummary` | (existing) |
| `checklists.*`, `agent.checklists.*` | `ChecklistService` | `checklistMutator` |

## Consequences

**Positive:**

- Correctness by construction. Adding a new mutation path inherits invariant maintenance for free; no convention to remember.
- Spoofing impossible. The wiki derives all attribution from auth context.
- Surfaces in tooling. `wiki-cli describe api.v1.<service>` documents the contract; `wiki-cli list` shows users the canonical entry points.
- Test discipline. The funnel has one place to assert metadata invariants, instead of N call sites.

**Negative:**

- More API surface. Each reserved namespace needs a service, breaking the convenience of "just use `MergeFrontmatter`."
- Migration cost when extending. Reserving a previously-loose namespace requires auditing every caller. (See the agent prompt audit work in #984.)
- Coupled to wiki's auth provider. The funnel derives attribution from auth context, so this works only inside the wiki process — external tools can't mutate reserved namespaces directly. (This is by design.)

**Neutral:**

- Generic frontmatter tools remain the right answer for loose data. Don't reserve a namespace unless there's a real invariant to defend.

## When to Apply This Pattern

Use the reserved-namespace + dedicated-service + funnel pattern when **any** of the following are true:

- The wiki must maintain server-stamped fields (timestamps, counters, sync tokens).
- The wiki must derive fields from auth context (attribution).
- The data has identity (UIDs, sort-stable ordering) that external writes must not corrupt.
- Downstream consumers (sync protocols, scheduled fires, indexes) depend on metadata advancing correctly on every mutation.

Do **not** apply this pattern for:

- Loose, user-edited config (`title`, `description`, blog/inventory metadata).
- Data the wiki only reads but never enforces invariants on.

## References

- `help_scheduled_agents` — original reserved-namespace policy for `agent.*`.
- #984 — checklist data model refactor (this ADR's home).
- #983 — CalDAV bridge (consumer of the funnel for sync correctness).
````

### PR DESCRIPTION
**Pull request by brendanjerwin**
## Summary

Implements #984: refactors the checklist data model so per-item attribution, optimistic concurrency, and the per-list `sync_token` advance correctly on every mutation. Hard prerequisite for the CalDAV bridge (#983).

The change establishes a **reserved `wiki.*` frontmatter namespace** for wiki-managed metadata (per ADR-0010), introduces a typed **`ChecklistService`** (gRPC + MCP + wiki-cli) with `AddItem`/`UpdateItem`/`ToggleItem`/`DeleteItem`/`ReorderItem`/`ListItems`/`GetChecklists`, and routes every mutation through a single internal **`checklistMutator`** funnel. The `agent.*` namespace policy is generalized into a small registry per ADR-0009.

## What's in the box

**Architecture (ADRs):**
- `0009-reserved-namespaces-and-dedicated-services.md` — pattern doc for any reserved namespace.
- `0010-wiki-namespace-for-wiki-managed-metadata.md` — introduces `wiki.*` as the home for wiki-managed metadata that isn't agent-specific.

**Identity subsystem:**
- `IdentityValue.IsAgent()` and `IdentityValue.Name()` for attribution.
- `--agent-tag` CLI flag wires Tailscale node tags into the resolver.
- New `x-wiki-is-agent` gRPC metadata header lets anonymous and Tailscale-authenticated callers self-claim agent status. wiki-cli sets it by default; `WIKI_CLI_HUMAN=1` opts out.

**Reserved-namespace registry:**
- `internal/grpc/api/v1/reserved.go` replaces the ad-hoc `agent.*` helpers in `frontmatter.go` with a small registry.
- `agent.*` and `wiki.*` are wholesale-rejected by `MergeFrontmatter`/`ReplaceFrontmatter`/`RemoveKeyAtPath`.
- `checklists.*` is **not** reserved — raw frontmatter writes still succeed but bypass attribution + sync_token bookkeeping.

**ChecklistService:**
- Proto at `api/proto/api/v1/checklist.proto` with `(api.v1.description)` and `(api.v1.example_request)` extensions on every RPC. The MCP tool listing is what agents read.
- Server handlers in `internal/grpc/api/v1/checklist.go`. Each delegates to the mutator and maps errors to gRPC codes.

**`checklistMutator` (the funnel):**
- New `server/checklistmutator/` package owning every mutation.
- Diff-on-write: per-item `created_at`/`updated_at`, `completed_at`/`completed_by` (cleared on toggle-off), `automated` derived from `IdentityValue.IsAgent()`.
- Per-list `sync_token` increments on every logical mutation; `Checklist.updated_at` is the OCC ETag.
- Tombstones on delete with 7-day TTL, GC'd lazily on next mutation/read.
- OCC: `expected_updated_at` mismatch returns `FailedPrecondition`.
- Per-page mutex; single page-store round-trip per mutation.
- Codec splits state across user-facing `checklists.<list>.*` and reserved `wiki.checklists.<list>.*`.

**wiki-cli:**
- New `checklist` subcommand: `list/add/toggle/update/delete/reorder`.
- Outgoing transport wraps `http.Client` with `agentHeaderTransport` so every wiki-cli call carries `x-wiki-is-agent: true` by default. `WIKI_CLI_HUMAN=1` opts out.

**Frontend (`<wiki-checklist>`):**
- `static/js/web-components/wiki-checklist.ts` swapped from `Frontmatter` to `ChecklistService` Connect-ES client.
- Each user action → single typed RPC; polling short-circuits when `Checklist.updated_at` is unchanged.
- OCC `FailedPrecondition` → refetch + retry once + 2-second "Edited concurrently — refreshed" toast.
- New metadata UI: "Done by …", description rendering, due-date indicator with overdue styling, `localStorage.wiki-checklist-debug=1` debug badges.
- E2E test (`e2e/tests/checklist-polling.spec.ts`) routes updated to `**/api.v1.ChecklistService/*`.

**Eager migration:**
- `migrations/eager/checklist_data_model_migration.go` mirrors `checklist_tag_syntax_migration.go` line-for-line.
- Walks every page, assigns ULIDs to items lacking `uid`, backfills `sort_order`, populates `wiki.checklists.<list>.items.<uid>.{created_at, updated_at}`, stamps `wiki.checklists.<list>.migrated_data_model = true`. Idempotent.
- Registered in `server/site.go` `startMigrationJobs()` alongside the tag-syntax migration.

**Help docs:**
- `internal/syspage/embedded/help_macro_checklist.md` rewritten — drops the read-modify-write footgun, documents `ChecklistService`, the two-tier schema, OCC, attribution.
- `help_scheduled_agents.md` cross-references the new ADRs.

**External agent prompts (on this workstation):**
- `~/agent/skills/{wiki-checklist, wiki-confirmations, project-template, weekly-dinner-planner, daily-reflection, simple-wiki-overseer}/SKILL.md` updated to use `ChecklistService` MCP tools.
- `~/agent/config/heartbeat.md` — `Frontmatter key paths` section replaced with `Checklist mutations — use the typed service`.

## Test plan
- [x] `devbox run lint:everything` is green (Go lint + revive + staticcheck + golangci-lint + frontend tsc + ESLint + Mocha + Storybook build).
- [x] `devbox run go:test` is green; the new `server/checklistmutator/...` and `migrations/eager/checklist_data_model_migration_test.go` cover every invariant in the ADR (created_at on add, updated_at on edit, completed_at/by on toggle, sync_token monotonic, OCC mismatch → FailedPrecondition, mutex serializes concurrent mutations, single round-trip per mutation, tombstone GC on read/write, registered `wiki.*` rejection).
- [x] `devbox run fe:test` is green; `wiki-checklist.test.ts` exercises every RPC path, OCC retry, polling short-circuit, metadata captions, debug-flag gating; `wiki-checklist-cursor.integration.test.ts` adds an OCC-refetch cursor-preservation case.
- [ ] **Deploy** to test instance and verify:
  - First boot post-upgrade runs `ChecklistDataModelMigrationScanJob` (look for "Checklist data-model migration started." log line and confirm pages get stamped).
  - `MergeFrontmatter` with `wiki.*` payload returns InvalidArgument naming `ChecklistService`.
  - `MergeFrontmatter` with `checklists.*` payload still succeeds (deliberate non-reservation of user data).
  - Toggle item via Web UI; confirm "Done by …" caption and `wiki.checklists.<list>.items.<uid>.completed_by` populated.
  - Open same checklist in two tabs and trigger a concurrent edit — second mutation surfaces the "Edited concurrently — refreshed" toast.
  - `wiki-cli checklist add foo bar "test"` succeeds; resulting item has `automated = true`. `WIKI_CLI_HUMAN=1 wiki-cli checklist add foo bar "test"` produces `automated = false`.
- [ ] **Post-deploy:** sweep deployed scheduled-agent prompts (`mcp__wiki__api_v1_SearchService_SearchContent` for `MergeFrontmatter checklists`) and rewrite each via `AgentMetadataService.UpsertSchedule`.
- [ ] **Post-merge follow-up:** file an issue for the survey data-model refactor (mirror of #984 against `wiki-survey.ts` / `surveys.*`).

Closes #984.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by brendanjerwin** [comment]
@claude SonarCloud is failing the Quality Gate at **68.1% Coverage on New Code** (required ≥80%) for #984. Please add Go test coverage for the new code in this PR — primarily:

- `server/checklistmutator/codec.go` (decodeChecklist, decodeItem, decodeLegacyItem, decodeTombstones, encodeChecklist, encodeItem, encodeTombstones, listChecklistNames, the readMap/readSlice/readInt64/readTimestamp/stringValue/boolValue/int64Value/stringSlice helpers).
- `server/checklistmutator/mutator.go` — uncovered branches around Clock/ULID interface alternatives, error paths in `readFrontMatter` (ErrPageNotFound, generic read error), `densifyAroundSortOrder`, edge cases of `pruneTombstones`, the alarm-payload set/clear branch in UpdateItem, ReorderItem with new_sort_order == current.
- `migrations/eager/checklist_data_model_migration.go` — the scan-job code path (DataDirExists/ListMDFiles/extractDataModelMigrationFrontmatter), `legacyHasAny`, `pageNeedsDataModelMigration` for the non-system-page edge cases, all four migration shapes covered by unit tests, `ensureMapInParent` / `readNestedMap` / `ensureNestedMap` / `int64FromMap` / `int64Fallback`.
- `internal/grpc/api/v1/reserved.go` — the `reservedNamespaceError("agent")` and `reservedNamespaceError("wiki")` paths, `reservedKeyOnPath` with non-key components.
- `internal/grpc/api/v1/checklist.go` handlers — currently covered for happy path; add tests for `errChecklistMutatorNotConfigured` and the `mapChecklistMutatorErr` switch (ErrItemNotFound / ErrListNotFound / ErrPageNotFound / generic err).
- `cmd/wiki-cli/checklist.go` and `cmd/wiki-cli/agent_header.go` — table-driven tests for the subcommands and `newAgentAwareHTTPClient`.
- `tailscale/identity.go` IsAgent / Name on `*Identity` and `Anonymous` (already mostly covered) and `tailscale/identity_resolver.go` `nodeIsAgent` edge cases.

Stay on the `feat/issue-984-checklist-data-model` branch. Do not change behavior — only add tests. Run `devbox run lint:everything` before pushing. Report which percentage coverage you reached when done.

File: server/checklistmutator/mutator.go [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+
+// Mutator is the single funnel for checklist mutations. Construct one per
+// process via New; share across handlers.
+type Mutator struct {
+	pages wikipage.PageReaderMutator
+	clock Clock
+	ulids ulid.Generator
+
+	mu     sync.Mutex
+	pageMu map[string]*sync.Mutex
```
[the remark this answers is not in the document]
Comment (brendanjerwin) [reply server/checklistmutator/mutator.go]: Fixed in a50bcf3b — Mutator.pageMu is now a sync.Map. Monotonic growth is bounded by ever-mutated-page count and documented in the struct field comment.

File: server/checklistmutator/mutator.go [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+func densifyAroundSortOrder(items []*apiv1.ChecklistItem, movedIdx int) {
+	target := items[movedIdx].SortOrder
+	for i, it := range items {
+		if i == movedIdx {
+			continue
+		}
+		if it.SortOrder == target {
+			it.SortOrder++
+		}
+	}
```
[the remark this answers is not in the document]
Comment (brendanjerwin) [reply server/checklistmutator/mutator.go:583]: Fixed in a50bcf3b — densifyAroundSortOrder now walks ascending and bumps each later item past the previous one, so cascades resolve in one pass.

File: static/js/web-components/wiki-checklist.ts [hunk 3]
```diff
@@ -493,16 +638,41 @@ export class WikiChecklist extends LitElement implements DragReorderHandler {
...
+          ...(expectedUpdatedAt && { expectedUpdatedAt }),
+        })),
+        previousItems,
+        'deleting checked items'
+      );
+      // Bail if a delete failed — don't keep firing requests against
+      // a list that's already in a bad state.
+      if (this.error) return;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (brendanjerwin) [reply static/js/web-components/wiki-checklist.ts:675]: Documented in a50bcf3b — sequential is required because each delete advances Checklist.updated_at; Promise.all would race on OCC. Bulk DeleteItems RPC is the right answer for big batches; not adding the API surface for a typically <10-item case.
